# Submissions at Hackers News
Data set source: [here](https://www.kaggle.com/hacker-news/hacker-news-posts)

Start by adding a title and writing a paragraph in a markdown cell introducing the project and the data set. The title and the introduction are tentative at this point, so don't spend too much time here — you can come back at the end of your work to refine them.

* Read the hacker_news.csv file in as a list of lists. Assign the result to the variable hn.
* Display the first five rows of hn.


In [1]:
from csv import reader
open_file = open('hacker_news.csv')
reader = reader(open_file)
hn = list(reader)
import pprint
print(hn[:5])
pprint.pprint(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]
[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',


* Extract the first row of data, and assign it to the variable headers.
* Remove the first row from hn.
* Display headers.
* Display the first five rows of hn to verify that you removed the header row properly.


In [2]:
headers = hn[:1]
hn = hn[1:]
print("Headers:", headers)
print("Table:", hn[:5])

Headers: [['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]
Table: [['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']

* Create three empty lists called ask_posts, show_posts, and other_posts.
* Loop through each row in hn.
    * Assign the title in each row to a variable named title.
        * Because the title column is the second column, you'll need to get the element at index 1 in each row.

* Implement the following steps:
    * If the lowercase version of title starts with ask hn, append the row to ask_posts.
    * Else if the lowercase version of title starts with show hn, append the row to show_posts.
    * Else append to other_posts.

* Check the number of posts in ask_posts, show_posts, and other_posts.

In [3]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


In [4]:
print(ask_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


In [5]:
print(show_posts[:5])

[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], ['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11'], ['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']]


Next, let's determine if ask posts or show posts receive more comments on average.

* Find the total number of comments in ask posts and assign it to total_ask_comments.
    * Set total_ask_comments to 0.
* Use a for loop to iterate over the ask posts.
    * Because the num_comments column is the fifth column in ask_posts, you'll need to get the element at index 4 in each row.
        * You'll also need to convert the value to an integer so that we can calculate the sum of all the comments.
        * Add this value to total_ask_comments.
* Compute the average number of comments on ask posts and assign it to avg_ask_comments.
* Print avg_ask_comments.
* Find the total number of comments in show posts and assign it to total_show_comments.
    * Set total_show_comments to 0.
* Use a for loop to iterate over the show posts.
    * Because the num_comments column is the fifth column in show_posts, you'll need to get the element at index 4 in each row.
        * You'll also need to convert the value to an integer so that we can calculate the sum of all the comments.
        * Add this value to total_show_comments.
* Compute the average number of comments on show posts and assign it to avg_show_comments.
* Print avg_show_comments.
* Do show posts or ask posts receive more comments on average? Write a markdown cell explaining your findings.

In [6]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [7]:
total_show_comments = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


## Findings:
Average comments number for ask posts is 14.038417431192661.

Average comments number for show posts is 10.31669535283993.

Ask posts has more comments in average than show posts.

* Import the datetime module as dt.
* Create an empty list and assign it to result_list. This will be a list of lists.
* Iterate over ask_posts and append to result_list a list with two elements:
    * The first element shall be the column created_at.
        * Because the created_at column is the seventh column in ask_posts, you'll need to get the element at index 6 in each row.
    * The second element shall be the number of comments of the post.
        * You'll also need to convert the value to an integer.
* Create two empty dictionaries called counts_by_hour and comments_by_hour.
* Loop through each row of result_list.
* Extract the hour from the date, which is the first element of the row.
* Use the datetime.strptime() method to parse the date and create a datetime object.
* Use the string we want to parse as the first argument and a string that specifies the format as the second argument.
    * Use the datetime.strftime() method to select just the hour from the datetime object.
    * If the hour isn't a key in counts_by_hour:
        * Create the key in counts_by_hour and set it equal to 1.
        * Create the key in comments_by_hour and set it equal to the comment number.
    * If the hour is already a key in counts_by_hour:
        * Increment the value in counts_by_hour by 1.
        * Increment the value in comments_by_hour by the comment number.

In [8]:
import datetime as dt
result_list = []

for row in ask_posts:
    created_at = row[6]
    n_comments = int(row[4])
    result_list.append([created_at, n_comments])

counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
    created_at = row[0]
    created_date = dt.datetime.strptime(created_at, date_format)
    hour = created_date.strftime("%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = int(row[1])
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += int(row[1])
    
print("Counts by hour:", counts_by_hour)
print("Comments by hour", comments_by_hour)
pprint.pprint(counts_by_hour)

Counts by hour: {'12': 73, '23': 68, '06': 44, '20': 80, '11': 58, '14': 107, '17': 100, '03': 54, '00': 55, '21': 109, '04': 47, '15': 116, '16': 108, '19': 110, '10': 59, '07': 34, '09': 45, '13': 85, '01': 60, '02': 58, '18': 109, '08': 48, '22': 71, '05': 46}
Comments by hour {'12': 687, '23': 543, '06': 397, '20': 1722, '11': 641, '14': 1416, '17': 1146, '03': 421, '00': 447, '21': 1745, '04': 337, '15': 4477, '16': 1814, '19': 1188, '10': 793, '07': 267, '09': 251, '13': 1253, '01': 683, '02': 1381, '18': 1439, '08': 492, '22': 479, '05': 464}
{'00': 55,
 '01': 60,
 '02': 58,
 '03': 54,
 '04': 47,
 '05': 46,
 '06': 44,
 '07': 34,
 '08': 48,
 '09': 45,
 '10': 59,
 '11': 58,
 '12': 73,
 '13': 85,
 '14': 107,
 '15': 116,
 '16': 108,
 '17': 100,
 '18': 109,
 '19': 110,
 '20': 80,
 '21': 109,
 '22': 71,
 '23': 68}


* Use the example above to calculate the average number of comments per post for posts created during each hour of the day.
* The result should be a list of lists in which the first element is the hour and the second element is the average number of comments per post. Assign the result to a variable named avg_by_hour. Display the results.

In [9]:
avg_by_hour = []
for hour in counts_by_hour:
    avg = comments_by_hour[hour] / counts_by_hour[hour]
    avg_by_hour.append([hour, avg])
print(avg_by_hour)

[['12', 9.41095890410959], ['23', 7.985294117647059], ['06', 9.022727272727273], ['20', 21.525], ['11', 11.051724137931034], ['14', 13.233644859813085], ['17', 11.46], ['03', 7.796296296296297], ['00', 8.127272727272727], ['21', 16.009174311926607], ['04', 7.170212765957447], ['15', 38.5948275862069], ['16', 16.796296296296298], ['19', 10.8], ['10', 13.440677966101696], ['07', 7.852941176470588], ['09', 5.5777777777777775], ['13', 14.741176470588234], ['01', 11.383333333333333], ['02', 23.810344827586206], ['18', 13.20183486238532], ['08', 10.25], ['22', 6.746478873239437], ['05', 10.08695652173913]]


* Create a list that equals avg_by_hour with swapped columns.
    * Create an empty list and assign it to swap_avg_by_hour.
    * Iterate over the rows of avg_by_hour and append to swap_avg_by_hour a list whose first element is the second element of the row, and whose second element is the first element of the row.
* Print swap_avg_by_hour.
* Use the sorted() function to sort swap_avg_by_hour in descending order. Since the first column of this list is the average number of comments, sorting the list will sort by the average number of comments.
    * Set the reverse argument to True, so that the highest value in the first column appears first in the list.
    * Assign the result to sorted_swap.
* Print the string "Top 5 Hours for Ask Posts Comments".
* Loop through each average and each hour (in this order) in the first five lists of sorted_swap.
* Use the str.format() method to print the hour and average in the following format: 15:00: 38.59 average comments per post.
    * To format the hours, use the datetime.strptime() constructor to return a datetime object and then use the strftime() method to specify the format of the time.
    * To format the average, you can use {:.2f} to indicate that just two decimal places should be used.
* Which hours should you create a post during to have a higher chance of receiving comments? Refer back to the documentation for the data set to convert the times to the time zone you live in. Write a markdown cell explaining your findings.

In [10]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
print(swap_avg_by_hour)
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print("Sorted swap:", sorted_swap[:5])

[[9.41095890410959, '12'], [7.985294117647059, '23'], [9.022727272727273, '06'], [21.525, '20'], [11.051724137931034, '11'], [13.233644859813085, '14'], [11.46, '17'], [7.796296296296297, '03'], [8.127272727272727, '00'], [16.009174311926607, '21'], [7.170212765957447, '04'], [38.5948275862069, '15'], [16.796296296296298, '16'], [10.8, '19'], [13.440677966101696, '10'], [7.852941176470588, '07'], [5.5777777777777775, '09'], [14.741176470588234, '13'], [11.383333333333333, '01'], [23.810344827586206, '02'], [13.20183486238532, '18'], [10.25, '08'], [6.746478873239437, '22'], [10.08695652173913, '05']]
Sorted swap: [[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21']]


In [11]:
print("Top 5 Hours for Ask Posts Comments") #4
for row in sorted_swap[:5]: #5
    hour = row[1]
    avg = float(row[0])
    hour_dt = dt.datetime.strptime(hour, "%H")
    clean_hour = hour_dt.strftime("%H:%M")
    clean_avg = "{:.2f}".format(avg)
    print(clean_hour+": ", clean_avg, "average comments per post")

Top 5 Hours for Ask Posts Comments
15:00:  38.59 average comments per post
02:00:  23.81 average comments per post
20:00:  21.52 average comments per post
16:00:  16.80 average comments per post
21:00:  16.01 average comments per post


* Which hours should you create a post during to have a higher chance of receiving comments? Refer back to the documentation for the data set to convert the times to the time zone you live in. Write a markdown cell explaining your findings.

In [12]:
for row in sorted_swap[:5]:
    hour = row[1]
    hour_dt = dt.datetime.strptime(hour, "%H")
    spain_hour = hour_dt + dt.timedelta(hours = 6)
    clean_hour = spain_hour.strftime("%H:%M")
    avg = float(row[0])
    clean_avg = "{:.2f}".format(avg)
    print(clean_hour+": ", clean_avg, "average comments per post")

21:00:  38.59 average comments per post
08:00:  23.81 average comments per post
02:00:  21.52 average comments per post
22:00:  16.80 average comments per post
03:00:  16.01 average comments per post


**Findings**: Hours outside working hours (at night or early in the morning) are the ones with the most comments.